core/drivers/crypto/crypto_api/acipher/ecc.c

```diff
@@ -0,0 +1,404 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2020 NXP
+ *
+ * Crypto ECC interface implementation to enable HW driver.
+ */
+#include <crypto/crypto_impl.h>
+#include <drvcrypt.h>
+#include <drvcrypt_acipher.h>
+#include <tee/tee_cryp_utl.h>
+#include <utee_defines.h>
+
+/*
+ * Returns the key size in bits for the given ECC curve
+ *
+ * @curve   ECC Curve ID
+ */
+static size_t get_ecc_keysize(uint32_t curve)
+{
+	switch (curve) {
+	case TEE_ECC_CURVE_NIST_P192:
+		return 192;
+
+	case TEE_ECC_CURVE_NIST_P224:
+		return 224;
+
+	case TEE_ECC_CURVE_NIST_P256:
+		return 256;
+
+	case TEE_ECC_CURVE_NIST_P384:
+		return 384;
+
+	case TEE_ECC_CURVE_NIST_P521:
+		/* Key size is 528 bits to be byte aligned */
+		return 528;
+
+	default:
+		return 0;
+	}
+}
+
+/*
+ * Verify if the cryptographic algorithm @algo is valid for
+ * the ECC curve
+ *
+ * @curve   ECC curve
+ * @algo    Cryptographic algorithm
+ */
+static bool algo_is_valid(uint32_t curve, uint32_t algo)
+{
+	uint8_t algo_op = 0;
+	uint8_t algo_id = 0;
+	uint8_t algo_curve = 0;
+
+	algo_op = TEE_ALG_GET_CLASS(algo);
+	algo_id = TEE_ALG_GET_MAIN_ALG(algo);
+	algo_curve = TEE_ALG_GET_DIGEST_HASH(algo);
+
+	/* Check first if the aglo operation and the algo id are correct */
+	if ((algo_op == TEE_OPERATION_ASYMMETRIC_SIGNATURE &&
+	     algo_id == TEE_MAIN_ALGO_ECDSA) ||
+	    (algo_op == TEE_OPERATION_KEY_DERIVATION &&
+	     algo_id == TEE_MAIN_ALGO_ECDH)) {
+		if (curve == algo_curve) {
+			CRYPTO_TRACE("Algo 0x%" PRIx32 " curve %d is valid",
+				     algo, curve);
+			return true;
+		}
+	}
+
+	CRYPTO_TRACE("Algo 0x%" PRIx32 " curve %d is not valid", algo, curve);
+
+	return false;
+}
+
+/*
+ * Free an ECC public key
+ *
+ * @key   Public Key
+ */
+static void ecc_free_public_key(struct ecc_public_key *key)
+{
+	struct drvcrypt_ecc *ecc = NULL;
+
+	if (key) {
+		ecc = drvcrypt_get_ops(CRYPTO_ECC);
+		if (ecc) {
+			CRYPTO_TRACE("ECC Public Key free");
+			ecc->free_publickey(key);
+		}
+	}
+}
+
+/*
+ * Generates an ECC keypair
+ *
+ * @key        Keypair
+ * @size_bits  Key size in bits
+ */
+static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
+				       size_t size_bits __maybe_unused)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct drvcrypt_ecc *ecc = NULL;
+
+	/* Check input parameters */
+	if (!key) {
+		CRYPTO_TRACE("Parameters error key is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ecc = drvcrypt_get_ops(CRYPTO_ECC);
+	if (ecc)
+		ret = ecc->gen_keypair(key, get_ecc_keysize(key->curve));
+
+	CRYPTO_TRACE("ECC Keypair (%zu bits) generate ret = 0x%" PRIx32,
+		     size_bits, ret);
```

Comment:
`size_bits` is not used, maybe worth not printing it out?

Reply:
It could be interesting to get the asked value for the debug in case the `get_ecc_keysize` is bugged.